### PR TITLE
Unify logic for pure and non-pure enums

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This is similar to what I suggested in https://github.com/babel/babel/pull/15467#issuecomment-1456009171: it unifies the code path for the "pure" and "non pure" versions, so that we avoid having two different but very similar cases.

Instead of always generating the variable declaration followed by the IIFE, which as you pointed out doesn't work, I took the opposite approach and always generate the declaration with the IIFE as the initializer. It has the unfortunate side effect that `let X = (function () {})(X || {})` is not valid (because `X` is in TDZ), so we must instead generate `let X = (function () {})({})` in this case.

### CODE REVIEW

File: packages/babel-plugin-transform-typescript/test/fixtures/enum/non-scoped/output.js [hunk 1]
```diff
@@ -3,7 +3,7 @@ var E = /*#__PURE__*/function (E) {
   E[E["y"] = 2] = "y";
   return E;
 }(E || {});
-E = /*#__PURE__*/function (E) {
+E = function (E) {
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-typescript/test/fixtures/enum/non-scoped/output.js]: This was not needed, terser doesn't remove used variables (E) anyway.

File: packages/babel-plugin-transform-typescript/test/fixtures/enum/scoped/output.js [hunk 2]
```diff
@@ -1,5 +1,6 @@
 {
   // Uses 'let' within a scope
-  let E;
-  (function (E) {})(E || (E = {}));
+  let E = /*#__PURE__*/function (E) {
+    return E;
+  }({});
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-typescript/test/fixtures/enum/scoped/output.js:5]: Now terser can remove this :)

File: packages/babel-plugin-transform-typescript/test/fixtures/enum/non-scoped/output.js (hunk 1 above)
Comment (liuxingbaoyu) [reply packages/babel-plugin-transform-typescript/test/fixtures/enum/non-scoped/output.js]: I'm inclined to keep it, maybe in the future terser can be implemented or other compressors can be used?

Comment (nicolo-ribaudo) [reply packages/babel-plugin-transform-typescript/test/fixtures/enum/non-scoped/output.js]: Done!

**Comment by liuxingbaoyu** [comment]
Thanks! It's all really good!

**Pull request merged**